client/src/utils.ts

```diff
@@ -0,0 +1,175 @@
+const xregexp = require('xregexp');
+const slash = require('slash');
+const fsPath = require('path');
+const languageTagRegex = require('ietf-language-tag-regex');
+
+const vscode = require('vscode')
+import { URI } from 'vscode-uri';
+
+// Matches 'sample.md' or 'sample_TAG.md' where TAG is an IETF language tag (BCP 47)
+const IETF_REGEXP = languageTagRegex({ exact: false }).toString().slice(1,-2);
+const SAMPLE_FILE_REGEXP = xregexp('text[/\\\\]sample(_(' + IETF_REGEXP + '))?.md$');
+
+
+function normalizeNLs(contents){
+        // we replace all \r and \n with \n
+        let text =  contents.replace(/\r/gm,'');
+        return text;
+}
+
+function registerFormulas(parserManager,logicManager){
+	const formulas = parserManager.getFormulas();
+	formulas.forEach(x => {
+		logicManager.addTemplateFile(x.code,x.name);
+	});
+}
+
+async function loadFilesContents(path, regex) {
+
+	const subdirs = await vscode.workspace.fs.readDirectory(URI.parse(path));
+	const result = await Promise.all(subdirs.map(async (subdir) => {
+	// subdir[1]=1
+	const res = fsPath.resolve(path, subdir[0]);
+
+		if(subdir[1]==2){
+			if( /.*node_modules$/.test(res) === false) {
+				return loadFilesContents(res, regex);
+			}
+			else {
+				return null;
+			}
+		}
+		else {
+			if(regex.test(res)) {
+				return {
+					name: res,
+					contents: await loadFileContents(path, res, false, true)
+				};
+			}
+			else {
+				return null;
+			}
+		}
+	}));
+	return result.reduce((a, f) => a.concat(f), []).filter((f) => f !== null);
+}
+
+async function loadFileBuffer(path, fileName, required=false) {
+	const filePath = fsPath.resolve(path, fileName);
+
+	try{
+		return await vscode.workspace.fs.readFile(URI.file(filePath));
+	}
+	catch(e) {
+		return null;
+	}
+	return null;
+}
+
+async function loadFileContents(path, fileName, json=false, required=false) {
+	const filePath = fsPath.resolve(path, fileName);
+
+	try {
+		const contents = Buffer.from(await vscode.workspace.fs.readFile(URI.file(filePath))).toString();
+		if(json && contents) {
+			return JSON.parse(contents);
```

Comment:
What if the caller thinks the file should be JSON but it isn't. JSON.parse will throw and exception here but it will be hard to debug what has gone wrong, because the exception stack trace is lost and the value returned in null.